### PR TITLE
1221 webhook to cx override 1

### PR DESCRIPTION
**Pull request by jonahkaye**
Ref: #1105 

### Upstream

- https://github.com/metriport/metriport/pull/1087/files. 

### Description

- add a secret flag called whOverride inside meta payload within /document/query and /consolidated/query/ routes that causes the webhook to not be sent to customer when called. Helpful for testing. 
- This PR has been branched from https://github.com/metriport/metriport/pull/1087/files. 

### Release Plan

- [ ] merge 
- [ ] test on staging
- [ ] release

### CODE REVIEW

File: docs/medical-api/more-info/webhooks.mdx [hunk 1]
```diff
@@ -96,6 +100,12 @@ Example payload:
       ...
     }
   ]
+  "documentRequestMetadata": {
+    "<some-request-meta-property>": "<the-property-value>"
+  },
+  "consolidatedRequestMetadata": {
+    "<some-request-meta-property>": "<the-property-value>"
+  },
```
Comment (Goncharo) [inline docs/medical-api/more-info/webhooks.mdx:108]: why not just put this in an optional `data` field in the webhook `meta` base model?

this sets a pattern for all WHs in our API, so not sure why we'd want a bunch of permutations like `documentRequestMetadata`, for each type - this seems to make it more cumbersome for devs

lmk what you think

File: docs/medical-api/api-reference/document/start-document-query.mdx [hunk 2]
```diff
@@ -37,6 +37,13 @@ If there's no document to be converted, the total will be set to zero and the st
...
 </ParamField>
 
+## Request Body
+
+<BodyField name="meta" type="object" optional>
+  A record of string key-value pairs. You can specify up to 50 keys, with key names up to 40 characters long and values up to 500 characters long. This metadata will be returned in the webhook response.
+</BodyField>
+
+
 ## Response
```
[the remark this answers is not in the document]
Comment (Goncharo) [reply docs/medical-api/api-reference/document/start-document-query.mdx:47]: also, maybe would suggest naming this `metadata`?

the `meta` sent in the actual WH will be a different shape, and could be confusing

**Review by Goncharo — CHANGES_REQUESTED** [review]
some comments for consideration - the main one is that it seems like including the metadata in a `data` property in the `meta` in the WH model would be a better desgin

lmk what you think, happy to chat about this

**Comment by leite08** [comment]
It seems this PR's branch was created from another, upstream branch. In this case, you should edit the PR so it points to the correct base branch:

image

This will make it so GH only shows the diff introduced in this PR, versus showing changes from this branch and the ones introduced on the upstream branch.

See how this confusing how Dima reviewed stuff related to the other PR on this one.

File: packages/api/src/command/medical/patient/consolidated-webhook.ts [hunk 3]
```diff
@@ -69,15 +68,16 @@ export const processConsolidatedDataWebhook = async ({
...
-    await processRequest(
-      webhookRequest,
-      settings,
-      bundle
-        ? {
-            bundleLength: optionalToString(bundle.entry?.length ?? bundle.total) ?? "unknown",
-          }
-        : undefined
-    );
+    if (!currentPatient.data.whOverride)
```
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-webhook.ts:71]: This prop name is not clear as to what it does. What about `disableWH`?

File: packages/api/src/command/medical/document/document-webhook.ts [hunk 4]
```diff
@@ -56,7 +56,7 @@ export const processPatientDocumentRequest = async (
     };
     const webhookRequest = await createWebhookRequest({ cxId, type: whType, payload });
     // send it to the customer and update the request status
-    await processRequest(webhookRequest, settings);
+    if (!patient.data.whOverride) await processRequest(webhookRequest, settings);
```
Comment (leite08) [inline packages/api/src/command/medical/document/document-webhook.ts:59]: We should cancel the flow before we create the WH request on our DB, right?

Here and on the consolidated one

File: packages/api/src/domain/medical/patient.ts [hunk 5]
```diff
@@ -70,6 +70,7 @@ export type PatientData = {
   externalData?: PatientExternalData;
   cxDocumentRequestMetadata?: unknown;
   cxConsolidatedRequestMetadata?: unknown;
+  whOverride?: boolean;
```
Comment (leite08) [inline packages/api/src/domain/medical/patient.ts:73]: I think we should store this on each one of the props for docQuery and consolidated, so triggering a doc query disabling WH doesn't mess w/ the cx's consolidated one.

Could just store inside the unknown property?

File: packages/api/src/command/medical/patient/consolidated-get.ts [hunk 6]
```diff
@@ -57,7 +57,10 @@ export async function startConsolidatedQuery({
     dob: patient.data.dob,
     genderAtBirth: patient.data.genderAtBirth,
     address: patient.data.address,
+    cxDocumentRequestMetadata: cxConsolidatedRequestMetadata,
     cxConsolidatedRequestMetadata: cxConsolidatedRequestMetadata,
+    whOverride:
+      (cxConsolidatedRequestMetadata as { whOverrideFlag?: string })?.whOverrideFlag !== undefined,
```
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-get.ts:63]: I suggest we create a simple function for this so:
1. its isolated from the rest of the code;
2. can be reused so we don't write this twice.

Also, it would be good to have the property/type/name defined in one place and referenced throughout, so we don't have the risk of updating it in one place and forgetting to update other places. By using `... as { whatever: string }` we're creating a new type that's disconnected from the one used somewhere else.

Actually, if we have this function inside `cxDocumentRequestMetadata` and `cxConsolidatedRequestMetadata` we should be able to simplify this PR to be:
- One function to check if `disableWH` is enabled
- Two conditionals calling that function

```ts
function isDisableWH(meta?: unknown): boolean {
  if (!meta) return false;
  return Boolean((meta as any).disableWH);
}
```

...used like:

```ts
  // send it to the customer and update the WH request status
  if (!isDisableWH(currentPatient.data.cxConsolidatedRequestMetadata)) {
    await processRequest(...);
  }
```

**Comment by leite08** [comment]
Also, please update the code pushed by [this PR](https://github.com/metriport/metriport/pull/1114), addressing the two entries `TODO 1106`.

File: packages/api/src/command/medical/document/document-webhook.ts (hunk 4 above)
Comment (jonahkaye) [reply packages/api/src/command/medical/document/document-webhook.ts:59]: No @leite08 the idea is that when we internally use the endpoint, we can access the results in the DB but the customer doesn't get the webhook request results.

File: packages/api/src/command/medical/patient/consolidated-webhook.ts (hunk 3 above)
Comment (jonahkaye) [reply packages/api/src/command/medical/patient/consolidated-webhook.ts:71]: Sgtm!

File: packages/api/src/command/medical/document/document-webhook.ts [hunk 7]
```diff
@@ -42,14 +44,19 @@ export const processPatientDocumentRequest = async (
...
-    const settings = await getSettingsOrFail({ id: cxId });
+    const [settings, patient] = await Promise.all([
+      getSettingsOrFail({ id: cxId }),
+      getPatientOrFail({ id: patientId, cxId }),
+    ]);
+
     // create a representation of this request and store on the DB
     const payload: WebhookPatientDataPayloadWithoutMessageId = {
       patients: [{ patientId, documents, status }],
+      documentRequestMetadata: patient.data.cxDocumentRequestMetadata,
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/api/src/command/medical/document/document-webhook.ts]: no a confusion from not having the right base branch sorry

File: packages/api/src/command/webhook/webhook.ts [hunk 8]
```diff
@@ -217,3 +217,8 @@ export const sendTestPayload = async (url: string, key: string): Promise<boolean
   if (res.pong && res.pong === ping) return true;
   return false;
 };
+
+export const isDisableWH = (meta?: unknown): boolean => {
```
Comment (Goncharo) [inline packages/api/src/command/webhook/webhook.ts:221]: nit: a better name could be `webhookIsDisabled`, will be easier to lookup in the future, and the caller doesn't need to know what `WH` means

File: packages/utils/src/commonwell/doc-query-shared.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  log?: typeof console.log;
+};
+
+export async function queryDocsForPatient({
+  cxId,
+  patientId,
+  apiUrl,
+  apiKey,
+  config = {
+    patientChunkDelayJitterMs: 1000,
```
Comment (Goncharo) [inline packages/utils/src/commonwell/doc-query-shared.ts:33]: what's all this stuff? doesn't look related

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: packages/utils/src/commonwell/doc-query-shared.ts (hunk 9 above)
Comment (jonahkaye) [reply packages/utils/src/commonwell/doc-query-shared.ts:33]: They are for TODO 1106 mentioned by @leite08 above.

File: packages/api/src/command/medical/document/document-webhook.ts [hunk 10]
```diff
@@ -54,12 +54,20 @@ export const processPatientDocumentRequest = async (
...
+        settings,
+        undefined,
+        patient.data.cxDocumentRequestMetadata
+      );
+    } else {
+      const status = "success";
+      await updateWebhookRequestStatus({
+        id: webhookRequest.id,
+        status,
+      });
```
Comment (leite08) [inline packages/api/src/command/medical/document/document-webhook.ts:69]: We talked about this on the huddle, we could create it already with the status "success", there's no need for an update right after a create, right?

File: packages/api/src/command/webhook/webhook.ts (hunk 8 above)
Comment (leite08) [reply packages/api/src/command/webhook/webhook.ts:221]: Agree that webhook is preferrable, but I suggest we keep `is` as prefix - has this on the previous team as a practice, boolean consts/variables start with `is` or `has`, it helps to quickly identify that's a boolean container.

File: packages/api/src/routes/medical/internal-docs.ts [hunk 11]
```diff
@@ -311,7 +312,7 @@ router.post(
  * @param req.query.cxId - The customer/account's ID.
  * @param req.query.patientId - The customer/account's ID.
  * @param req.query.facilityId - Optional; The facility providing NPI for the document query.
- *
+ * @param req.body Optional metadata to be sent through Webhook. WH disable can be sent here.
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-docs.ts:315]: > WH disable can be sent here.

It would be better to use the exact property name, "WH disable" doesn't tell me what I should send on the payload.

File: packages/utils/src/commonwell/doc-query-shared.ts (hunk 9 above)
Comment (leite08) [reply packages/utils/src/commonwell/doc-query-shared.ts:33]: I don't understand... [this file is already on develop](https://github.com/metriport/metriport/blob/develop/packages/utils/src/commonwell/doc-query-shared.ts), why is it showing up as an addition? Didn't you notice that? Is your branch/branches synchronized with `develop` (`git rebase develop`)?

File: packages/utils/src/commonwell/doc-query-shared.ts [hunk 12]
```diff
@@ -58,10 +56,15 @@ export async function queryDocsForPatient({
   const metriportAPI = new MetriportMedicalApi(apiKey, {
     baseAddress: apiUrl,
   });
+
+  const metadata = {
+    metadata: {
+      disableWHFlag: "true",
```
Comment (leite08) [inline packages/utils/src/commonwell/doc-query-shared.ts:62]: Why are we always disabling it? There was a parameter for the exact reason to pass this from the code calling this function. The TODO was talking about the directive, not the parameter, as we also talked about in the huddle.

The code is not compiling, try to compile `package/utils`. I don't know why it was not spotted by CI, looking into it.

**Review by leite08 — CHANGES_REQUESTED** [review]

File: packages/api/src/command/medical/document/document-webhook.ts (hunk 10 above)
Comment (jonahkaye) [reply packages/api/src/command/medical/document/document-webhook.ts:69]: I thought we said the opposite but will change.

**Review by leite08 — APPROVED** [review]

**Review by Goncharo — APPROVED** [review]
